src/SonarQube.Client.Tests/Services/SonarQubeServiceTests.cs

```diff
@@ -1,20 +1,313 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
 using FluentAssertions;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+using SonarQube.Client.Models;
 
 namespace SonarQube.Client.Services.Tests
 {
     [TestClass]
     public class SonarQubeServiceTests
     {
         [TestMethod]
-        public void Ctor_WithNullClietn_ThrowsArgumentNullException()
+        public async Task ConnectAsync_WhenAlreadyConnected_ThrowsInvalidOperation()
+        {
+            // Arrange
+            var client = new Mock<ISonarQubeClient>();
+            client.Setup(x => x.ValidateCredentialsAsync(It.IsAny<ConnectionDTO>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(() => Result<CredentialsDTO>.Ok(new CredentialsDTO { AreValid = true }));
+            client.Setup(x => x.GetVersionAsync(It.IsAny<ConnectionDTO>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(() => Result<VersionDTO>.Ok(new VersionDTO { Version = "5.6" }));
+            var service = new SonarQubeService(client.Object);
+            await service.ConnectAsync(new ConnectionInformation(new Uri("http://mysq.com")), CancellationToken.None);
+
+            // Act
+            Func<Task> func = async () =>
+                await service.ConnectAsync(new ConnectionInformation(new Uri("http://mysq.com")), CancellationToken.None);
+
+            // Assert
+            client.VerifyAll();
+            func.ShouldThrow<InvalidOperationException>().WithMessage("This operation expects the service not to be connected.");
+        }
+
+        [TestMethod]
+        public void Ctor_DefaultValues()
+        {
+            // Arrange
+            var client = new Mock<ISonarQubeClient>();
+
+            // Act
+            var service = new SonarQubeService(client.Object);
+
+            // Assert
+            service.OrganizationsFeatureMinimalVersion.Should().Be(new Version(6, 2));
+        }
+
+        [TestMethod]
+        public void Ctor_WithNullClient_ThrowsArgumentNullException()
         {
             // Arrange & Act
             Action action = () => new SonarQubeService(null);
 
             // Assert
             action.ShouldThrow<ArgumentNullException>();
         }
+        [TestMethod]
+        public async Task EnsureIsConnected_WhenConnected_ShouldDoNothing()
```

Comment:
I'd keep `EnsureIsConnected` private as it's an implementation detail and not part of a service API (correct me if I'm wrong). The fact this method throws or not doesn't show us much. It's the fact that is used by others that matters and that's what we should be testing.

I'd test the other service methods if they throw when not connected.

Reply:
Without moving to another test fwk this is too boring (too much boilerplate) too test all cases hence this direct test. Note that I didn't put it `public` for testing purpose but just because I expect people to use the class through the interface and because the method is `Pure`.

Reply:
It is boring, but there is only about 5 methods, so it cannot be that bad... I can write that test if you want.

If you only expect the class only used via interface then this method is weird - it is public but we do not expect anybody to call it - why not make it private then?

Reply:
I've written the test in my branch and found `GetRoslynExportProfileAsync` does not invoke `EnsureIsConnected` - which shows this test has value.